doctr/models/detection/core.py

```diff
@@ -82,6 +83,108 @@ def bitmap_to_boxes(
+    def nms(boxes: np.ndarray, thresh: float = .5) -> List[int]:
+        """Perform non-max suppression, borrowed from <https://github.com/rbgirshick/fast-rcnn>`_.
+
+        Args:
+            boxes: np array of straight boxes: (*, 5)
+            thresh: iou threshold to perform box suppression.
+
+        Returns:
+            A list of box indexes to keep
+        """
+        x1 = boxes[:, 0]
+        y1 = boxes[:, 1]
+        x2 = boxes[:, 2]
+        y2 = boxes[:, 3]
+        scores = boxes[:, 4]
+
+        areas = (x2 - x1 + 1) * (y2 - y1 + 1)
+        order = scores.argsort()[::-1]
+
+        keep = []
+        while order.size > 0:
+            i = order[0]
+            keep.append(i)
+            xx1 = np.maximum(x1[i], x1[order[1:]])
+            yy1 = np.maximum(y1[i], y1[order[1:]])
+            xx2 = np.minimum(x2[i], x2[order[1:]])
+            yy2 = np.minimum(y2[i], y2[order[1:]])
+
+            w = np.maximum(0.0, xx2 - xx1 + 1)
+            h = np.maximum(0.0, yy2 - yy1 + 1)
+            inter = w * h
+            ovr = inter / (areas[i] + areas[order[1:]] - inter)
+
+            inds = np.where(ovr <= thresh)[0]
+            order = order[inds + 1]
+        return keep
+
+    def filter_boxes(
+        self,
+        boxes: np.ndarray,
+        nms_thresh: float = .8,
+        merging_thresh: float = .2,
+    ) -> np.ndarray:
+        """Filter an array of boxes: boxes with an iou > merging threshold will be merged, and
+        nms is performed with a nms threshold
+
+        Args:
+            boxes: array of boxes, shape (N, 5) or (N, 6)
+            nms_thresh: iou threshold to perform nms
+            merging_thresh: threshold used to merge boxes
+
+        Returns:
+            A np array of filtered boxes, containing at most N boxes.
+        """
+        # Array of rotated boxes: not supported for now
+        # TODO: support for rotated boxes (both NMS & merging algorithm)
+        if boxes.shape[1] == 6:
+            return boxes
+
+        if not np.any(boxes):
+            # If array of boxes is empty, return it
+            return boxes
+
+        # First perform NMS on array of boxes
+        keep = self.nms(boxes, thresh=nms_thresh)
+
+        # Update box list
+        box_list = [list(box) for i, box in enumerate(list(boxes)) if i in keep]
+
+        # Compute iou beween boxes
+        boxes = np.asarray(box_list)
+        iou_mat = box_iou(boxes[:, :4], boxes[:, :4])
+        tri_mat = np.tril(iou_mat, -1)
+
+        def enclosing(box_a: List[float], box_b: List[float]) -> List[float]:
+            """Compute enclosing bbox from 2 list of 5 floats [xmin, ymin, xmax, ymax, score].
+            Return a list of [xmin, ymin, xmax, ymax, score].
+            """
+            x1, y1, x2, y2 = box_a[:4]
+            x3, y3, x4, y4 = box_b[:4]
+            score = (box_a[4] + box_b[4]) / 2
+            enclosing = [min(x1, x3), min(y1, y3), max(x2, x4), max(y2, y4), score]
+            return enclosing
```

Comment:
Might be better to define it as a static class method or move it to `doctr.utils.geometry` as a function